src/rez/pip.py

```diff
@@ -81,35 +86,106 @@ def find_pip(pip_version=None, python_version=None):
-        py_exe, pip_version, context = find_pip_from_context(
-            python_version,
-            pip_version=pip_version or "latest"
-        )
+    for version in [pip_version, pip_version or "latest"]:
+        try:
+            py_exe, found_pip_version, context = find_pip_from_context(
+                python_version, pip_version=version
+            )
+            if _check_found(py_exe, found_pip_version):
+                break
+        except BuildError as error:
+            print_warning(str(error))
 
-    if not py_exe:
+    else:  # No break from: for version in [pip_version,...]
         import pip
-        pip_version = pip.__version__
+
+        found_pip_version = pip.__version__
         py_exe = sys.executable
-        print_warning(
-            "Found no pip in python and pip package; "
-            "falling back to pip installed in rez own virtualenv (version %s)",
-            pip_version
+        print_warning("Found no pip in any python and/or pip packages!")
+        print_warning("Falling back to pip installed in rez own virtualenv:")
+        logging_arguments = (
+            ("pip", found_pip_version, pip.__file__),
+            ("python", ".".join(map(str, sys.version_info[:3])), py_exe),
         )
+        for warn_args in logging_arguments:
+            print_warning("%10s: %s (%s)", *warn_args)
 
-    pip_major = pip_version.split('.')[0]
-    if int(pip_major) < 19:
-        raise RezSystemError("pip >= 19 is required! Please update your pip.")
+    if not _check_found(py_exe, found_pip_version, log_invalid=False):
+        message = "pip{specifier} is required! Please update your pip."
+        raise RezSystemError(message.format(specifier=PIP_SPECIFIER))
 
     return py_exe, context
 
 
+def find_python_from_package(context, name="python", default=None):
+    """Get Python executable for current resolved, python/pip context.
+
+    See nerdvegas/rez#826.
+
+    1. Get resolved python package from ResolvedContext
+    2. Create a new context with just that python package, no system paths.
+    3. Do a `which python` to get py_exe
+
+          For windows: just try "python"
+          For *NIX: try in order "python#.#", "python#", "python"
+
+    Args:
+        context (ResolvedContext): Current context with resolved Python.
+        name (str): Name of the package for Python instead of "python".
+        default (str): Force a particular fallback path for Python executable.
+
+    Returns:
+        str or None: Path to Python executable, if any.
+    """
+    from rez.shells import create_shell
+
+    py_exe_path = default
+
+    # 1.
+    py_packages = (v for v in context.resolved_packages if v.name == name)
+    python_package = next(py_packages)
+
+    # Filthy hack to not have system paths added in ResovledContext._execute
+    # by RexExecutor.append_system_paths(), which calls Shell.get_syspaths()
+    shell_class = create_shell().__class__
+    current_syspaths = shell_class.syspaths
+    try:
```

Comment:
Yeah I don't think we should do this. Not because it's a hack per se, but because it isn't threadsafe. Ie this could alter syspaths in a context being created in another thread.